File: src/data/writtenQuestion.test.ts
import { test, describe } from 'node:test'
import assert from 'node:assert/strict'
import type { ManagedContentItem } from './contentControl.ts'
import { parseWrittenParts } from './questionFormat.ts'
import {
  managedWrittenToStudentWritten, markWritten, writtenFullyMarked, writtenPartsInOrder,
} from './writtenQuestion.ts'

const PARTS = parseWrittenParts(`### (a) 5 marks
Enumerate the contents of the femoral triangle.
Expects: Femoral nerve
Expects: Femoral artery
Expects: Femoral vein
Expects: Deep inguinal lymph nodes
Expects: Femoral canal

### (b) 4 marks
Summarise the ligaments of the hip joint.
Expects: Iliofemoral
Expects: Pubofemoral
Expects: Ischiofemoral
Expects: Ligamentum teres`)

function written(overrides: Partial<ManagedContentItem> = {}): ManagedContentItem {
  return {
    id: 'Q-W1',
    kind: 'question',
    title: 'Femoral triangle and hip joint',
    subjectId: 'msk',
    status: 'Published',
    owner: 'omar',
    updatedAt: '2026-08-21T00:00:00.000Z',
    fields: { Topic: 'Lower limb' },
    questionData: {
      format: 'structured_written',
      writtenParts: PARTS,
      attachments: [], correctAnswer: 'A', answers: [], attachedImage: '',
      libraryIds: [], resourceIds: [],
      tags: {
        topic: 'Lower limb', mainConceptIds: ['CON-MSK-1', 'CON-MSK-2'],
        conceptIds: [], contextualConceptIds: ['CON-MSK-9'],
      } as never,
      learningObjective: 'Name the femoral triangle contents and hip ligaments.',
      authorNotes: '', sourceCitation: '', estimatedSeconds: 600, randomiseAnswers: false,
    },
    ...overrides,
  } as ManagedContentItem
}

describe('Projecting a written question for a student', () => {
  test('the parts, their marks and the total come through', () => {
    const question = managedWrittenToStudentWritten(written())!
    assert.equal(question.parts.length, 2)
    assert.equal(question.totalMarks, 9)
    assert.equal(question.topic, 'Lower limb')
    assert.equal(question.parts[0].expectedPoints.length, 5)
  })

  test('contextual concepts are left out, as they are for an MCQ', () => {
    const question = managedWrittenToStudentWritten(written())!
    assert.deepEqual(question.conceptIds, ['CON-MSK-1', 'CON-MSK-2'])
    assert.ok(!question.conceptIds!.includes('CON-MSK-9'),
      'a concept the question mentions but never tests must not be offered as revision')
  })

  test('an unpublished written question is not shown', () => {
    assert.equal(managedWrittenToStudentWritten(written({ status: 'Draft' })), null)
  })

  test('an MCQ is not a written question', () => {
    const mcq = written()
    mcq.questionData!.format = 'mcq_single_best'
    assert.equal(managedWrittenToStudentWritten(mcq), null)
  })

  test('a written question with no parts is refused', () => {
    const empty = written()
    empty.questionData!.writtenParts = []
    assert.equal(managedWrittenToStudentWritten(empty), null,
      'there would be nothing to mark against, which is the whole of the practice')
  })
})

describe('Marking a written answer', () => {
  test('an unmarked answer scores null, not zero', () => {
    assert.equal(markWritten(null, PARTS), null,
      'a student who has written an answer and not marked it has not scored nothing')
  })

  test('marks are apportioned across a part\'s expected points', () => {
    const score = markWritten({
      'part-a': ['Femoral nerve', 'Femoral artery', 'Femoral vein', 'Femoral canal', 'Deep inguinal lymph nodes'],
      'part-b': ['Iliofemoral', 'Pubofemoral'],
    }, PARTS)!
    assert.equal(score.parts[0].marks, 5, 'all five of five points on a 5-mark part')
    assert.equal(score.parts[1].marks, 2, 'two of four points on a 4-mark part')
    assert.equal(score.marks, 7)
    assert.equal(score.outOf, 9)
  })

  test('a tick for a point that no longer exists does not count', () => {
    // Editing a question must not leave an old tick earning marks for a point
    // that was removed.
    const score = markWritten({ 'part-a': ['Femoral nerve', 'A point since deleted'] }, PARTS)!
    assert.equal(score.parts[0].covered, 1)
    assert.equal(score.parts[0].marks, 1)
  })

  test('a part with no published mark scheme scores nothing rather than everything', () => {
    const unmarkable = parseWrittenParts('### (a) 6 marks\nDiscuss the coronary circulation.')
    const score = markWritten({ 'part-a': [] }, unmarkable)!
    assert.equal(score.parts[0].marks, 0)
    assert.equal(score.parts[0].total, 0)
    assert.equal(score.outOf, 6, 'the paper still says six marks were available')
  })

  test('the parts total exactly what the paper carries', () => {
    // Rounding each part and then totalling drifts from the mark on the page.
    const score = markWritten({ 'part-a': ['Femoral nerve'], 'part-b': ['Iliofemoral'] }, PARTS)!
    assert.equal(score.parts[0].marks + score.parts[1].marks, score.marks)
    assert.equal(score.marks, 1 + 1)
  })
})

describe('Knowing when marking is finished', () => {
  test('every part with a scheme must be marked', () => {
    assert.equal(writtenFullyMarked({ 'part-a': [] }, PARTS), false)
    assert.equal(writtenFullyMarked({ 'part-a': [], 'part-b': [] }, PARTS), true)
    assert.equal(writtenFullyMarked(null, PARTS), false)
  })

  test('a part with nothing to tick does not strand the student', () => {
    const mixed = parseWrittenParts(`### (a) 2 marks
Answerable.
Expects: Something

### (b) 3 marks
No scheme was ever published for this.`)
    assert.equal(writtenFullyMarked({ 'part-a': ['Something'] }, mixed), true)
  })
})

describe('The order parts are answered in', () => {
  test('a part comes after the one it depends on', () => {
    const parts = parseWrittenParts(`### (b) 3 marks
Second.
Depends on: a

### (a) 2 marks
First.`)
    assert.deepEqual(writtenPartsInOrder(parts).map((part) => part.label), ['a', 'b'])
  })

  test('independent parts keep the order the paper printed them in', () => {
    assert.deepEqual(writtenPartsInOrder(PARTS).map((part) => part.label), ['a', 'b'])
  })

  test('a dependency cycle falls back to the printed order rather than hanging', () => {
    const parts = parseWrittenParts(`### (a) 1 marks
One.
Depends on: b

### (b) 1 marks
Two.
Depends on: a`)
    const ordered = writtenPartsInOrder(parts)
    assert.equal(ordered.length, 2)
    assert.deepEqual(new Set(ordered.map((part) => part.label)), new Set(['a', 'b']))
  })
})

File: android/app/src/test/java/com/synapse/app/core/essays/WrittenQuestionTest.kt
package com.synapse.app.core.essays

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class WrittenQuestionTest {

    private val partA = WrittenPart(
        id = "p-a", label = "a", prompt = "Enumerate the contents of the femoral triangle.",
        marks = 4, expectedPoints = listOf("Femoral nerve", "Femoral artery", "Femoral vein", "Lymphatics"),
    )
    private val partB = WrittenPart(
        id = "p-b", label = "b", prompt = "Summarise the ligaments of the hip joint.",
        marks = 6, expectedPoints = listOf("Iliofemoral", "Pubofemoral", "Ischiofemoral"),
    )

    // --- markWritten ------------------------------------------------------------

    @Test fun markWrittenIsNullWhenNotYetMarked() {
        assertNull(markWritten(null, listOf(partA, partB)))
    }

    @Test fun markWrittenApportionsMarksAcrossExpectedPointsPerPart() {
        val ticks = mapOf("p-a" to listOf("Femoral nerve", "Femoral artery"), "p-b" to listOf("Iliofemoral"))

        val score = markWritten(ticks, listOf(partA, partB))!!

        assertEquals(2, score.parts.size)
        assertEquals(2, score.parts[0].covered)
        assertEquals(4, score.parts[0].total)
        assertEquals(2.0, score.parts[0].marks, 0.0001) // 2/4 of 4 marks
        assertEquals(1, score.parts[1].covered)
        assertEquals(3, score.parts[1].total)
        assertEquals(2.0, score.parts[1].marks, 0.0001) // 1/3 of 6 marks
        assertEquals(4.0, score.marks, 0.0001)
        assertEquals(10, score.outOf)
    }

    @Test fun markWrittenIntersectsTicksWithPointsThatStillExist() {
        val ticks = mapOf("p-a" to listOf("Femoral nerve", "A stale point that no longer exists"))

        val score = markWritten(ticks, listOf(partA))!!

        assertEquals(1, score.parts.single().covered)
    }

    @Test fun markWrittenScoresZeroForAPartWithNoPublishedScheme() {
        val unmarkablePart = partA.copy(expectedPoints = emptyList())
        val ticks = mapOf("p-a" to emptyList<String>())

        val score = markWritten(ticks, listOf(unmarkablePart))!!

        assertEquals(0.0, score.parts.single().marks, 0.0001)
        assertEquals(4, score.parts.single().outOf)
    }

    // --- writtenFullyMarked -------------------------------------------------------

    @Test fun writtenFullyMarkedIsFalseWhenNotMarkedAtAll() {
        assertFalse(writtenFullyMarked(null, listOf(partA, partB)))
    }

    @Test fun writtenFullyMarkedRequiresEveryMarkablePartToBeTicked() {
        val onlyA = mapOf("p-a" to listOf("Femoral nerve"))
        assertFalse(writtenFullyMarked(onlyA, listOf(partA, partB)))

        val both = mapOf("p-a" to listOf("Femoral nerve"), "p-b" to emptyList<String>())
        assertTrue(writtenFullyMarked(both, listOf(partA, partB)))
    }

    @Test fun writtenFullyMarkedTreatsAPartWithNoExpectedPointsAsDone() {
        val unmarkablePart = partA.copy(expectedPoints = emptyList())
        assertTrue(writtenFullyMarked(emptyMap(), listOf(unmarkablePart)))
    }

    // --- WrittenProjection --------------------------------------------------------

    private val publishedWritten = """
        { "id":"W1","kind":"question","title":"Femoral triangle and hip ligaments","subjectId":"SYS_MSK","status":"Published",
          "fields":{"Topic":"Lower limb"},
          "questionData":{"format":"structured_written","learningObjective":"Recall regional anatomy.",
            "writtenParts":[
              {"id":"p-a","label":"a","prompt":"Enumerate the contents of the femoral triangle.","marks":4,
                "expectedPoints":["Femoral nerve","Femoral artery","Femoral vein","Lymphatics"],"conceptIds":["C1"]},
              {"id":"p-b","label":"b","prompt":"Summarise the ligaments of the hip joint.","marks":6,
                "expectedPoints":["Iliofemoral","Pubofemoral","Ischiofemoral"],"conceptIds":["C2"],"dependsOnPartId":"p-a"}
            ],
            "tags":{"mainConceptIds":["C1","C2"],"conceptIds":["C1","C2"]}}}
    """.trimIndent()

    // status != Published -> dropped.
    private val draftWritten = """
        { "id":"W2","kind":"question","title":"Draft written","subjectId":"SYS_MSK","status":"Draft",
          "questionData":{"format":"essay","writtenParts":[{"id":"p-a","label":"a","prompt":"x","marks":5,"expectedPoints":["x"]}]}}
    """.trimIndent()

    // format not in WRITTEN_FORMATS (single-best MCQ) -> dropped, even with writtenParts present.
    private val nonWrittenFormat = """
        { "id":"W3","kind":"question","title":"An MCQ","subjectId":"SYS_MSK","status":"Published",
          "questionData":{"format":"mcq_single_best","writtenParts":[{"id":"p-a","label":"a","prompt":"x","marks":5,"expectedPoints":["x"]}]}}
    """.trimIndent()

    // No parts -> nothing to mark against -> dropped.
    private val noPartsWritten = """
        { "id":"W4","kind":"question","title":"No parts","subjectId":"SYS_MSK","status":"Published",
          "questionData":{"format":"short_answer","writtenParts":[]}}
    """.trimIndent()

    // kind != "question" -> ignored entirely.
    private val nonQuestionItem = """
        { "id":"E1","kind":"essay","title":"Not a written question","subjectId":"SYS_MSK","status":"Published",
          "essayData":{"prompt":"x","keyPoints":[{"id":"kp-0","text":"x"}]}}
    """.trimIndent()

    @Test fun projectsOnlyThePublishedMarkableWrittenQuestion() {
        val ledgerJson = "[$publishedWritten,$draftWritten,$nonWrittenFormat,$noPartsWritten,$nonQuestionItem]"

        val questions = WrittenProjection.project(ledgerJson)

        assertEquals(1, questions.size)
        val q = questions.single()
        assertEquals("W1", q.id)
        assertEquals("Femoral triangle and hip ligaments", q.title)
        assertEquals("SYS_MSK", q.subjectId)
        assertEquals("Lower limb", q.topic)
        assertEquals("Femoral triangle and hip ligaments", q.stem)
        assertEquals(2, q.parts.size)
        assertEquals(10, q.totalMarks)
        assertEquals("Recall regional anatomy.", q.learningObjective)
        assertEquals(listOf("C1", "C2"), q.conceptIds)
        assertEquals("p-a", q.parts[0].id)
        assertEquals(4, q.parts[0].marks)
        assertEquals("p-a", q.parts[1].dependsOnPartId)
    }

    @Test fun acceptsLedgerWrappedInItemsObject() {
        val ledgerJson = """{ "items": [$publishedWritten] }"""

        val questions = WrittenProjection.project(ledgerJson)

        assertEquals(1, questions.size)
        assertEquals("W1", questions.single().id)
    }
}
